packages/apollo-engine-reporting/src/schemaReporter.ts

```diff
@@ -0,0 +1,166 @@
+    if (schemaReporter.stopped()) return;
+
+    // Not awaiting this. The callback is handled in the `then` and it calls inner()
+    // to report the server info in however many seconds we were told to wait from
+    // Apollo Graph Manager
+    schemaReporter
+      .reportServerInfo(sendNextWithExecutableSchema)
+      .then(({ inSeconds, withExecutableSchema }) => {
+        sendNextWithExecutableSchema = withExecutableSchema;
+        setTimeout(inner, inSeconds * 1000);
+      })
+      .catch((error: any) => {
+        // In the case of an error we want to continue looping
+        // We can add hardcoded backoff in the future,
+        // or on repeated failures stop responding reporting.
+        logger.error(
+          `Error in reporting server info to Apollo Graph Manager for schema reporting: ${error}`,
+        );
+        sendNextWithExecutableSchema = false;
+        setTimeout(inner, fallbackReportingDelayInMs);
+      });
+  }
+
+  inner();
+}
+
+interface ReportServerInfoReturnVal {
+  inSeconds: number;
+  withExecutableSchema: boolean;
+}
+
+// This class is meant to be a thin shim around the gql mutations.
+export class SchemaReporter {
+  // These mirror the gql variables
+  private readonly serverInfo: EdgeServerInfo;
+  private readonly executableSchemaDocument: any;
+  private readonly url: string;
+
+  private isStopped: boolean;
+  private readonly headers: Headers;
+
+  constructor(
+    serverInfo: EdgeServerInfo,
+    schemaSdl: string,
+    apiKey: string,
+    schemaReportingEndpoint: string | undefined,
+  ) {
+    this.headers = new Headers();
+    this.headers.set('Content-Type', 'application/json');
+    this.headers.set('x-api-key', apiKey);
+    this.headers.set('apollographql-client-name', 'apollo-engine-reporting');
+    this.headers.set(
+      'apollographql-client-version',
+      require('../package.json').version,
+    );
+
+    this.url =
+      schemaReportingEndpoint ||
+      'https://engine-graphql.apollographql.com/api/graphql';
+
+    this.serverInfo = serverInfo;
+    this.executableSchemaDocument = schemaSdl;
+    this.isStopped = false;
+  }
+
+  public stopped(): Boolean {
+    return this.isStopped;
+  }
+
+  public stop() {
+    this.isStopped = true;
+  }
+
+  public async reportServerInfo(
+    withExecutableSchema: boolean,
+  ): Promise<ReportServerInfoReturnVal> {
+    const { data, errors } = await this.graphManagerQuery({
+      info: this.serverInfo,
+      executableSchema: withExecutableSchema
+        ? this.executableSchemaDocument
+        : null,
+    } as ReportServerInfoVariables);
+
+    if (errors) {
+      throw new Error((errors || []).map((x: any) => x.message).join('\n'));
+    }
+
+    if (!data || !data.me || !data.me.__typename) {
+      throw new Error(`
+Heartbeat response error. Received incomplete data from Apollo Graph Manager.
+If this continues please reach out at support@apollographql.com.
+Got response: "${JSON.stringify(data)}"
+      `);
```

Comment:
I find the prefix a bit repetitive here but I think we can also add some specificity to where this error occurred, which can be more important when it's happening asynchronously on a timer, since it will lack a more clear stack trace in some cases.

Also, perhaps more importantly, it's worth noting that using template literals (`` ` ``) results in the preservation of newlines.  Personally, I've found those less ergonomic to be present in `Error.prototype.message` properties since they are often collapsed to single lines or surface in reporting pipelines where the secondary/tertiary lines can be missed or blend in with other log lines.  This is amplified in this case by the first line of the error message being blank in this case.  That brings another concern about error reporting pipelines which naively preserve only the first line as the error message and assume that the subsequent lines are from the stack traces.

My first suggestion here merely changes the text of the error, but my second suggestion further quashes the newlines — curious to get feedback on my perspective on this though:

How about?:

### Just changed error message

```suggestion
      throw new Error(`
Unexpected response shape from Apollo Graph Manager when reporting server information during schema registration.
If this continues, please reach out to support@apollographql.com.
Got response: "${JSON.stringify(data)}"
      `);
```

### Also changing newline positioning.

```suggestion
      throw new Error([
        "Unexpected response shape from Apollo Graph Manager when",
        "reporting server information during schema registration.  If",
        "this continues, please reach out to support@apollographql.com.",
        "Received response: ",
        JSON.stringify(data)",
      ].join(" "));
```